### DISCUSSION TITLE
Automatic Static Optimization on custom server

### DISCUSSION THREAD

**Discussion by zwhitchcox** [vercel/next.js#17563] [Ideas]
# Feature request
Automatic Static Optimization on custom server

## Is your feature request related to a problem? Please describe.

I need to do several things like manage my own database on the server, interact with files on the server directly, upload images, initialize my database on the server, etc., and my only option is to create a custom server.

## Describe the solution you'd like

I don't understand why the automatic static page optimization can't be done on a custom server. If you're just trying to get me to use the Vercel platform, that loses a lot of good will from me. It seems like that would be pretty easy to port whatever code you use on your server to the `next` function.

## Describe alternatives you've considered

Adding a million dependencies to my app and signing up for a million services.

**Comment by timneutkens** [comment]
Overall we recommend not adding a custom server, not to make you use Vercel but to make sure we can optimize the whole stack end to end. Automatic static optimization is always there, but if you're using a custom server there's some downsides like you can't remap routes which can lead to bugs in your application, hence why we don't recommend it.

**Reply by AbdelrahmanTorki** [reply]
I hope you simplify middleware functionality such as express.js implementation. Unfortunately, current middleware isn't simply to implement and has many limitations such as: 

- Passing data from middleware to the API. 
- Edge runtime errors due to the lack of support for some Node.js modules, making many other packages unusable.
- The api file system. The page route to file wrapping is awesome, but in case of the API, it's distracting in development. You end up using multiple files and nesting for the development, even though it may look logical, it's distracting. It's easier to use one file for all related api routes .e.g. _user.js_ for user related routes, _product.js_ for product related routes. It's much easier than using "user/login", "user/register", "user/../..etc.".
- Finally, middleware chaining ease of implementation and websocket. 

Those are the main issues I think people tend to use custom servers such as express.js, nest.js. If such issues weren't there, next.js would be self sufficient.

**Comment by zwhitchcox** [comment]
So, you're saying it's impossible to add that stuff to the custom server?

**Reply by timneutkens** [reply]
I'm saying that `next()` is the same as `next start` but booted programmatically. There's just a bunch of gotchas like you can't remap urls that are automatic static optimization.

**Reply by zwhitchcox** [reply]
Ohhhh ok, so the pages are still optimized statically, but just the routes that are custom are not static, because next doesn't control them?

**Reply by timneutkens** [reply]
> but just the routes that are custom are not static

Not exactly, they'd serve the static file but breaks the router in the browser as we can't match the url to hydrate the router. Hence why I'm saying there's some gotchas.

**Reply by zwhitchcox** [reply]
Ok, I'm not using it for routing anyway...thanks. My mistake, I should've given you guys more credit.

And in fact I think I've found a way I might be able to use Vercel...thanks for your help!

**Reply by gaurav5430** [reply]
> > but just the routes that are custom are not static
> 
> Not exactly, they'd serve the static file but breaks the router in the browser as we can't match the url to hydrate the router. Hence why I'm saying there's some gotchas.

can you elaborate this? What do you mean by remapping urls and matching the url on the browser?

**Reply by jonsamp** [reply]
I assume what's meant by remapping urls is that with a custom server, you can map a url like `/@[username]` to the next page `/profile`. That particular use-case is important for us and a big reason why we need a custom server, since next pages don't allow characters like `@` in dynamic page names. (`/[username]` works, but `/@[username]` does not)

Re-mapping urls to next pages can cause issues with how pages download chunks, and cause unnecessary full page reloads, which can make a site slower. I faced some of those issues, but we were able to resolve them.

Im still wondering: How can we enable automatic static optimization using a custom server? Is there a protocol or some specification about how to set up a custom server to allow it? Or are we only able to get these optimization features by using Vercel's service?
